OTP FIPS mode fixes

npmccallum opened the pull request:
The following patches add various fixes for using OTP on a server running in FIPS mode.

npmccallum commented:
@frozencemetery FYI

frozencemetery commented:
Thank you for including the log message :)

Tested a variety of tokens in FIPS mode.  Please note: this will work for all new OTPs.  It will also work for existing OTPs provided that they were not sha512 or sha384 (the default is sha1, so this probably won't affect many users).

rcritten commented on daemons/ipa-slapi-plugins/libotp/hotp.c:
```diff
@@ -66,6 +67,49 @@ static const struct {
+                                     PK11_ATTR_PRIVATE |
+                                     PK11_ATTR_SENSITIVE, NULL);
+    if (!ekey)
+        goto egress;
```
There should be a log message so we know it failed here along with the NSS error code.

npmccallum replied:
The NSS error code is useless (SECFailure). You have to use GDB. This is also unlikely to fail except for OOM.

rcritten replied:
PR_GetError()

npmccallum replied:
It appears we can't log from this code. See the missing symbol error in the CI. I'm going to remove the patch which adds logging (at least temporarily).

rcritten commented on daemons/ipa-slapi-plugins/libotp/hotp.c:
```diff
@@ -66,6 +67,49 @@ static const struct {
+    /* Encrypt the input key. */
+    if (PK11_Encrypt(ekey, CKM_AES_CBC_PAD, &ivitem, ctitem.data, &ctitem.len,
+                     ctitem.len, key->data, key->len) != SECSuccess)
+        goto egress;
```
There should be a log message so we know it failed here along with the NSS error code.

npmccallum replied:
The NSS error code is useless (SECFailure). You have to use GDB. This is also unlikely to fail except for OOM.

rcritten commented on daemons/ipa-slapi-plugins/libotp/hotp.c:
```diff
@@ -66,6 +67,49 @@ static const struct {
+    /* Unwrap the input key. */
+    skey = PK11_UnwrapSymKey(ekey, CKM_AES_CBC_PAD, &ivitem,
+                             &ctitem, mech, CKA_SIGN, key->len);
+
```
What happens if the unwrap fails?

npmccallum replied:
Then you break out GDB.

rcritten replied:
Shouldn't it at least be checked?  static analyzers are sure to whing.

npmccallum replied:
I'm not sure what the problem is here. If an error occurs in `import_key()`, we return `NULL`. This is true for all code paths. The `skey` variable is initialized to `NULL`, so static analyzers shouldn't complain.

rcritten commented on daemons/ipa-slapi-plugins/libotp/hotp.c:
```diff
@@ -66,6 +67,49 @@ static const struct {
+                             &ctitem, mech, CKA_SIGN, key->len);
+
+egress:
+    PK11_FreeSymKey(ekey);
```
If in the case where creating the encryption key fails will PK11_FreeSymKey() blow up if ekey is NULL?

npmccallum replied:
```
180	PK11_FreeSymKey(PK11SymKey *symKey)
181	{
182	    PK11SlotInfo *slot;
183	    PRBool freeit = PR_TRUE;
184	
185	    if (!symKey) {
186	        return;
187	    }
```

rcritten replied:
nice.

mkosek commented:
@rcritten, this is the Bugzilla with the context of this work:
https://bugzilla.redhat.com/show_bug.cgi?id=1544679
Merging Nathaniel's work should enable us to also revert the changes that were done in https://pagure.io/freeipa/issue/7168

tiran commented:
CI is failing:

```
ipa_kdb.c:639:13: warning: ‘ipadb_free’ defined but not used [-Wunused-function]
 static void ipadb_free(krb5_context context, void *ptr)
             ^~~~~~~~~~
ipa_kdb.c:634:14: warning: ‘ipadb_alloc’ defined but not used [-Wunused-function]
 static void *ipadb_alloc(krb5_context context, void *ptr, size_t size)
              ^~~~~~~~~~~
```

tiran commented:
No, that was just a red herring. The actual error is:

```
./.libs/libhotp.a(hotp.o): In function `hmac':
/freeipa/rpmbuild/BUILD/freeipa-4.6.90.dev201802221933+gitca42e8e/daemons/ipa-slapi-plugins/libotp/hotp.c:146: undefined reference to `slapi_log_error'
collect2: error: ld returned 1 exit status
```

pvoborni commented:
https://pagure.io/freeipa/issue/7243 seems to me like a good fit for these patches.

npmccallum commented:
I've pushed a new set of patches, including reverting https://pagure.io/freeipa/issue/7168. It is very unfortunate that we can't log in this code path. But perhaps @frozencemetery can tackle this in a separate patch?

frozencemetery commented:
Is logging a blocker for merging this?  (Asking for my own prioritization.)

npmccallum commented:
IMHO, no.

npmccallum commented:
I pushed a new version. It only changes the text wrapping on the commit messages.

rcritten commented:
IMHO yes. The urgency is now so the patch should be made complete now. Pushing things off is usually equivalent to never doing them.

The problem if any of these fail is that we'll have no idea what at all has happened, things will just not work. We won't necessarily have a reproduction so what, we (or GSS) try to walk the user through a gdb session? Why would we even try that? What would point in that direction? If there is a log somewhere saying some call failed along with the NSS return code we at least get an inkling of where it is failing and hopefully why.

The final call, skey =..., never has the return value checked. So skey could be NULL and only the caller would find out. Is the caller expecting a NULL?

npmccallum commented:
Regarding the unchecked `skey`, yes the caller expects to receive NULL. See:
https://github.com/freeipa/freeipa/pull/1621/commits/a47b66a4ada7c28c59eb7085c40bbd621141389f#diff-21d729f4715e316ae0f969b07fb0ecbaR138

npmccallum commented:
Regarding logging, I see only one call that is likely to fail. I added logging for it, but then realized that this code doesn't actually link against 389ds and thus can't log (see the CI failure at the top of this PR). I removed that patch from this PR since it caused CI failures. All of the other calls are extremely likely to succeed and no logging exists for the existing crypto code in this file. Thus, I consider such logging an enhancement.

If FreeIPA maintainers disagree, then we need to do some more invasive changes to allow this code to support logging.

rcritten commented:
Can it log to syslog? That should be relatively straightforward. I know I have a difficult time debugging OTP in general and this is a black box without gdb.

npmccallum commented:
I honestly don't know the answer to that. Maybe @frozencemetery knows?

frozencemetery commented:
"Yes", it can.  It's a bit gross though - there's no state, so we never call `openlog()/closelog()`.  Messages end up in syslog like: `mar 01 14:15:05 freeipa.dbsplat.biz ns-slapd[1843]: message text`.

See the top commit on my branch over here: https://github.com/frozencemetery/freeipa/tree/logging

mkosek commented:
The added logs by @frozencemetery look promising. But let me CC here @tbordaz in case he has any advice for the FreeIPA DS plugin logging.

tbordaz commented:
I agree that currently libotp can not log. One possibility could be to change an libotp init function (otp_config_init ?) to pass a callback that will call slapi_log_error. Then libotp will use this callback.

frozencemetery commented:
Since this seems to be blocked on logging, what approach does the project want to see?  The raw `syslog()` I implemented?  API change?  Something else?

frozencemetery commented:
Bump - how does logging need to work for this to be merged?

rcritten commented:
Your syslog patch works for me.

I'm inclined to ack, will wait for others to comment before acking and pushing COB Mar 7.

frozencemetery commented:
Sorry to keep bumping, but can this be pushed, then?

tbordaz commented:
Using syslog() is good to me and avoid doing a complex fix to provide a logging function.
So I also ack this option

rcritten commented:
master:

* c9c58f2d35dfb8a6d1e187d260013812dcfdfd90 Fix OTP validation in FIPS mode
* a01a24ce5aded934095ea428d539ca30850f72e3 Increase the default token key size
* d498d7272d8de3e24afc442a3b001518fd98ebff Revert "Don't allow OTP or RADIUS in FIPS mode"

rcritten commented:
A manual backport to the ipa-4-5 branch is going to be required.
